contrib/mesos/pkg/executor/executor.go

```diff
@@ -447,308 +426,148 @@ func (k *Executor) attemptSuicide(driver bindings.ExecutorDriver, abort <-chan s
+		k.sendStatus(driver, newStatus(task.TaskId, mesos.TaskState_TASK_FAILED, err.Error()))
 		return
 	}
 
-	//TODO(jdef) check for duplicate pod name, if found send TASK_ERROR
-
-	// send the new pod to the kubelet which will spin it up
-	task.podName = podFullName
-	k.pods[podFullName] = pod
-	ok := k.sendPodsSnapshot()
-	if !ok {
-		task.podName = ""
-		delete(k.pods, podFullName)
-		return // executor is terminating, cancel launch
+	err = k.registry.bind(task.TaskId.GetValue(), pod)
+	if err != nil {
+		log.Errorf("failed to bind task %v pod %v/%v: %v",
+			task.TaskId.GetValue(), pod.Namespace, pod.Name, err)
+		k.sendStatus(driver, newStatus(task.TaskId, mesos.TaskState_TASK_FAILED, err.Error()))
+		return
 	}
 
-	// From here on, we need to delete containers associated with the task upon
-	// it going into a terminal state.
-
-	// report task is starting to scheduler
-	statusUpdate := &mesos.TaskStatus{
-		TaskId:  mutil.NewTaskID(taskId),
+	// send TASK_STARTING
+	k.sendStatus(driver, &mesos.TaskStatus{
+		TaskId:  task.TaskId,
 		State:   mesos.TaskState_TASK_STARTING.Enum(),
 		Message: proto.String(messages.CreateBindingSuccess),
-		Data:    data,
-	}
-	k.sendStatus(driver, statusUpdate)
-
-	// Delay reporting 'task running' until container is up.
-	psf := podStatusFunc(func() (*api.PodStatus, error) {
-		return k.podStatusFunc(pod)
+		Data:    startingData,
 	})
-	go k._launchTask(driver, taskId, podFullName, psf, task.launchTimer.C)
-}
 
-func (k *Executor) _launchTask(driver bindings.ExecutorDriver, taskId, podFullName string, psf podStatusFunc, expired <-chan time.Time) {
-	getMarshalledInfo := func() (data []byte, cancel bool) {
-		// potentially long call..
-		if podStatus, err := psf(); err == nil && podStatus != nil {
-			select {
-			case <-expired:
-				cancel = true
-			default:
-				k.lock.Lock()
-				defer k.lock.Unlock()
-				if _, found := k.tasks[taskId]; !found {
-					// don't bother with the pod status if the task is already gone
-					cancel = true
-					break
-				} else if podStatus.Phase != api.PodRunning {
-					// avoid sending back a running status before it's really running
-					break
-				}
-				log.V(2).Infof("Found pod status: '%v'", podStatus)
-				result := api.PodStatusResult{
-					ObjectMeta: api.ObjectMeta{
-						Name:     podFullName,
-						SelfLink: "/podstatusresult",
-					},
-					Status: *podStatus,
-				}
-				if data, err = json.Marshal(result); err != nil {
-					log.Errorf("failed to marshal pod status result: %v", err)
+	// within the launch timeout window we should see a pod-task update via the registry.
+	// if we see a Running update then we need to generate a TASK_RUNNING status update for mesos.
+	handlerFinished := make(chan struct{})
+	expired := make(chan struct{})
+	go func() {
+		defer close(expired)
+		<-launchTimer.C
+	}()
+	handler := podWatchHandler{
+		action: func(pod *RegisteredPod) (bool, error) {
+			if pod == nil {
```

Comment:
This `pod==nil` case feels strange and is not documented for podWatchHandler. Why not have an `expired` calllback next to `action`?